pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+on the form and shape of data, i.e. the structure of data objects.
+Nonetheless, we find that Python code is often sparkled with
+conditions depending on ``isinstance``, ``len``, ``getattr``,
+``hasattr``, etc.  Despite the benefits of 'duck typing', actual
+code still requires to query the format or type of an object in order
+to choose an appropriate processing action or to extract relevant bits
+of information.
+
+Unfortunately, the conditions for slightly more complex structures
+quickly grow into sequences of partly interdependent bits of
+structural tests, hurting readbility and maintability.  Pattern
+matching offers here a more direct scheme of expressing the minimal
+structure that an objects needs to have in order to allow for further
+processing.  Rather than writing a series of manual tests, patterns
+follow a *declarative* style, which is well-known for improved
+readability, maintability, and for delegating the burden of efficient
+execution on the compiler and interpreter.
+
+The concept of pattern matching is similar to regular expressions,
+where succinct patterns describe a textual structure.  A dedicated
+compiler then transforms these declarative patterns into highly
+efficient finite state machines.  In contrast to regular expressions,
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
+match.  Moreover, pattern matching blends the matching of a suitable
+pattern with that of a function, i.e. code that is executed in order
+to handle and process the information provided by a specific kind of
+object.
+
+One of the simplest forms of pattern matching that we find in other
+languages comes in the form of *function overloading*.  The type and
+number of arguments then determine which implementation of a specific
+function will be executed.  Object-oriented languages (including
+Python) may also use the *visitor pattern* to differentiate an action
+based on the type or class of an object.  Both of these approaches,
+however, are aimed at 'shallow' structures with little or no direct
+support for nested structures or structural information that is not
+directly encoded in an object's class or type.  For instance, it is
+simple to differentiate between an integer, a string, and a tuple, say,
+but becomes quite cumbersome and difficult to differentiate between
+tuples of different lengths, or between one containing string elements
+vs. one containing numeric elements.  This is where pattern matching
+shines: for structures that go beyond simple class distinctions.
+
+Although pattern matching is a concept that has been known and used
+for decades, we propose a re-interpretation that centres around the
+principle of 'duck typing' and builds on existing features of the
+Python language such as iterable unpacking.  Patterns adopt the syntax
+of parameters as far as possible and, to a somewhat lesser degree,
```

Comment:
"as far as possible"? I could argue that we might have gone further... :-) I'd rather stick to sequence/iterable unpacking as a model.

Or is this anticipating the rule that we only bind unqualified names? Honestly that's less important to me than the analogy to unpacking. (In a world where we marked either loads or stores we could capture to dotted names just fine, but pattern matching without sequence unpacking just wouldn't be the same.)

Reply:
The load/store semantics is only one (small) aspect.  As mentioned above, a successful case clause typically binds values to a couple of local variables and then executes some code.  This really is the basic job description of a function.  Moreover, parameters in a function have a strong _declarative_ nature.  When looking at `def foo(a, b)`, say, the question in what order `a` and `b` are bound is kind of nonsensical and moot.

Iterable unpacking has a strong temporal component to it: the assignments are performed in a well defined order, which means that `x, x = 2, 3` has a well-defined behaviour, whereas you would/could not write `def foo(x, x)`.  And, of course, iterable unpacking then sparks all this discussion about lvalues, which I don't think are a very productive way of approach it.

On the other hand, I understand what you mean with the unpacking being a central piece of the game and I agree with you.  I was mentally thinking that Python has unpacking in function calls, too, anyway (e.g. `foo(*seq)`), but that is a mistake on my part, because it applied to calls rather than the declaration.  So, since Python 3 removed unpacking from formal parameters, this aspect is indeed missing in my approach.  I will have to ponder that for a bit...